### ISSUE TITLE
[HxButton] Cannot specify HxTooltip span CSS classes

### ISSUE THREAD

**Issue by robertmclaws** [havit/Havit.Blazor#17]
I have a button group I'm implementing where I need to be able to set the padding on the button content. However, when a Tooltip is specified, the HxButton.CssClass is not set on the `<span />` used to contain the tooltip, so it renders incorrectly.

![image](https://user-images.githubusercontent.com/1657085/135740793-2fc631d4-2e2b-47e1-a54e-cb4fa78b8f73.png)
The one on the top is the proper rendering, using the original controls I'm replacing. the one on the bottom is with HxButtons. The first button does not have a tooltip specified, whereas the other 2 do. So `px-2` is set on the `<button />`. not the `<span />` that contains it.

**Comment by hakenr** [comment]
Hi @robertmclaws, we are aware of this issue and investigating the possibilities to fix.
Current implementation of HxTooltip (also used in HxButton) does not render the span at all if the `HxTooltip.Text` is not set.
The was introduced to make the HxButton as lightweight as possible (pure `button` element) if there is no `HxButton.Tooltip` set (which we believe is the by far most common case).
The current implementation has to be changed, but still not sure in which direction:
1. drop `HxButton.Tooltip` support
2. always render the HxTooltip's `span` (and set the `HxButton.CssClass` to it)

Can you help us with arguments for one or another direction? Or any third one?

**Comment by robertmclaws** [comment]
I think the current implementation is fine, TBH. I liked the thoughtfulness of how it worked. I believe my PR handled it efficiently in a way that allows you to specify CSS classes on both the `<button>` and the `<span>` predictably.

I can see an argument for always rendering the span, though that might bother people trying to achieve 1:1 HTML ports.

I think the documentation just needs to be clearer inside the Button Tooltip demo that adding a tooltip will wrap the button in a span.

**Comment by hakenr** [comment]
We have to fix this together with the implementation of HxTooltip. The current behavior is weird, if you do not set `HxTooltip.Text` nothing gets rendered except the child content. What if you set `HxTooltip.SpanCssClass` and leave the `HxTooltip.Text` empty? Do you expect the `SpanCssClass` not being used at all?

**Comment by robertmclaws** [comment]
It depends on whether or not you render the span all the time. Not rendering the container if the Text is empty makes a lot of sense, so I would keep that behavior.

I wanted to break the PR for this into two parts. If you first agree with the HxTooltip fixes, which were pretty straightforward, the next step would be to add TooltipContainerCssClass and TooltipCssClass properties to HxButton, to pass those values into the HxTooltip control.

Honestly, I think the current implementation is fine. Aside from being surprised about the container span, and not having control over the its properties, its very well designed! These fixes plus clearer docs will solve those problems quite nicely.

**Comment by hakenr** [comment]
OK, fine, I would just like to have the naming for `HxButton.Tooltip` and `HxTooltip` aligned. The `HxButton.TooltipContainerCssClass` and `HxTooltip.ContainerCssClass`  (instead of `SpanCssClass`) works better for me.

**Comment by robertmclaws** [comment]
Are you implementing this in a PR, or would you like me to?

**Comment by hakenr** [comment]
Now the first part is prepared, the HxTooltip - In the end I decided for `HxTooltip.WrapperCssClass` and to render the wrapper not only if the `Text` is set but also when the `WrapperCssClass` is set.

**Comment by hakenr** [comment]
@robertmclaws
I will implement the HxButton changes, I'm just not sure whether the `HxButton.CssClass` should automatically be rendered to the tooltip's `WrapperCssClass` if the `HxButton.Tooltip` is set, or whether the `CssClass` should always stay with the `button` and the `TooltipWrapperCssClass` will be separate parameter.

There might be a rare case where the `HxButton.Tooltip` will be set dynamically (dynamically rendered UI or something like that). Do we want the users to have to decide where to put their classes in such cases?

```cshtml
 <HxButton
       Tooltip="@dynamicTooltip"
       CssClass="@(String.IsNullOrEmpty(dynamicTooltip) ? "my-class" : null)"
       TooltipWrapperCssClass="@(String.IsNullOrEmpty(dynamicTooltip) ? null : "my-class")"
       ... />
```

...the answer might be in the fact that I made the tooltip's `span` rendered whenever the `WrapperCssClass` is set (no matter of the `HxTooltip.Text` value) => which might lead to a simple instruction for users: _Use `TooltipWrapperCssClass` whenever you intend to use `HxButton.Tooltip` no matter there is a chance the tooltip remains unset._

But isn't it too confusing? Should we use just one `HxButton.CssClass` and place it to the `span`/`button` as needed? Is it important for the users to know whether the class will be set to the span or button? Is there any scenario where they will want to set both? 

**Ha! Is this your scenario where you need to set both `<HxButton ... CssClass="btn-light-5" TooltipWrapperCssClass="px-2" />`?**

**Comment by robertmclaws** [comment]
> Ha! Is this your scenario where you need to set both <HxButton ... CssClass="btn-light-5" TooltipWrapperCssClass="px-2" />?

Yep, I'm using it in dynamic UI right now (actually ran into an issue where the OnClick handler is not bindable, so I had to change how the buttons load in order to make it work). That scenario is exactly what I need.

This is a really powerful library, so I think you should err on the side of predictability, and don't overthink it. I think you give the developer the ability to control how it renders, and if that means they have to have their own version of the button's GetButtonCssClass() in their code, then so be it.

> _Use TooltipWrapperCssClass whenever you intend to use HxButton.Tooltip no matter there is a chance the tooltip remains unset._

Absolutely, telling them this makes it predictable, I don't think you should be fancy or try to outsmart the end user. TooltipCssClass and TooltipWrapperCssClass are super intuitive and predictable, no need to do anything else.

**Comment by hakenr** [comment]
Resolved in release 1.3.7.
